### PR TITLE
feat(aca-py): oob receive and create invitation

### LINKED ISSUE #211: Support out of band api in ACA-Py
**Describe your new idea or enhancement**

I would like to use the out of band API from acapy with the agent-cli

### PR DESCRIPTION
**Pull request by morrieinmaas**
closes #211

### CODE REVIEW

**Review by blu3beri — CHANGES_REQUESTED** [review]

File: crates/agent/src/modules/oob.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+
+/// Response from the cloudagent when an invitation is created
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct OobConnectionCreateInvitationResponse {
+    /// the invitation message id generated by the cloudagent
+    /// this will be used for further functionality
+    pub invi_msg_id: String,
```
[the remark this answers is not in the document]
Comment (morrieinmaas) [reply crates/agent/src/modules/oob.rs:12]: we get this back in the payload - eg:
```json
{
  "created_at": "2021-12-31T23:59:59Z",
  "invi_msg_id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "invitation": {
    "@id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "@type": "https://didcomm.org/my-family/1.0/my-message-type",
    "handshake_protocols": [
      "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/didexchange/1.0"
    ],
    "label": "Bob",
    "requests~attach": [
      {
        "@id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "byte_count": 1234,
        "data": {
          "base64": "ey4uLn0=",
          "json": {
            "sample": "content"
          },
          "jws": {
            "header": {
              "kid": "did:sov:LjgpST2rjsoxYegQDRm7EL#keys-4"
            },
            "protected": "ey4uLn0",
            "signature": "ey4uLn0",
            "signatures": [
              {
                "header": {
                  "kid": "did:sov:LjgpST2rjsoxYegQDRm7EL#keys-4"
                },
                "protected": "ey4uLn0",
                "signature": "ey4uLn0"
              }
            ]
          },
          "links": [
            "https://link.to/data"
          ],
          "sha256": "617a48c7c8afe0521efdc03e5bb0ad9e655893e6b4b51f0e794d70fba132aacb"
        },
        "description": "view from doorway, facing east, with lights off",
        "filename": "IMG1092348.png",
        "lastmod_time": "2021-12-31T23:59:59Z",
        "mime-type": "image/png"
      }
    ],
    "services": [
      {
        "did": "WgWxqztrNooG92RXvxSTWv",
        "id": "string",
        "recipientKeys": [
          "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
        ],
        "routingKeys": [
          "did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH"
        ],
        "serviceEndpoint": "http://192.168.56.101:8020",
        "type": "string"
      },
      "did:sov:WgWxqztrNooG92RXvxSTWv"
    ]
  },
  "invitation_id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "invitation_url": "https://example.com/endpoint?c_i=eyJAdHlwZSI6ICIuLi4iLCAiLi4uIjogIi4uLiJ9XX0=",
  "oob_id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "state": "await_response",
  "trace": true,
  "updated_at": "2021-12-31T23:59:59Z"
}
```

[the remark this answers is not in the document]
Comment (morrieinmaas) [reply crates/agent/src/modules/oob.rs:12]: what would you rename it to?

File: crates/cli/src/modules/oob.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        .get(0)
+        .ok_or(Error::InvalidAgentInvitation)?;
+
+    // Base64 decode the invitation to a Vec<u8>
+    let decoded =
+        base64::decode(serialized_invitation).map_err(|_| Error::InvalidAgentInvitation)?;
+
+    // Convert the vec to a valid string
+    let decoded_str = str::from_utf8(&decoded)?;
+    print!("{decoded_str}");
```
[the remark this answers is not in the document]
Comment (morrieinmaas) [reply crates/cli/src/modules/oob.rs:142]: oops

[the remark this answers is not in the document]
Comment (morrieinmaas) [reply crates/cli/src/modules/oob.rs:142]: advanced debug ;P

File: crates/agent/src/modules/oob.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    /// Id of the request
+    pub request_id: Option<String>,
+
+    /// Id of the invitation message
+    pub invitation_msg_id: Option<String>,
+}
+
+/// Options supplied by the frontend for creating an invitation
+#[derive(Debug, Default)]
+pub struct OobConnectionCreateInvitationOptions {
```
[the remark this answers is not in the document]
Comment (morrieinmaas) [reply crates/agent/src/modules/oob.rs:81]: ??

[the remark this answers is not in the document]
Comment (blu3beri) [reply crates/agent/src/modules/oob.rs:82]: This structure is specific to the `cli` package and not the `agent`. I think we do this everywhere but I will have to refactor this out.

[the remark this answers is not in the document]
Comment (morrieinmaas) [reply crates/agent/src/modules/oob.rs:82]: ah ok. Then I can take it out here already

[the remark this answers is not in the document]
Comment (blu3beri) [reply crates/agent/src/modules/oob.rs:82]: Ah id rather just do everything at once. You can leave it here.

[the remark this answers is not in the document]
Comment (morrieinmaas) [reply crates/agent/src/modules/oob.rs:82]: alright then - keep your secrets...

can we merge then?

**Review by blu3beri — APPROVED** [review]

**Pull request merged**